### PR TITLE
Update Derivation Implementation

### PR DESCRIPTION
**Pull request by adamgfraser**
Apparently the documentation is not accurate. 😃

### CODE REVIEW

**Review by regiskuckaertz — APPROVED** [review]
😆

**Review by sideeffffect — APPROVED** [review]
Beautiful :heart:

File: core/shared/src/main/scala-3/zio/prelude/EqualVersionSpecific.scala [hunk 1]
```diff
@@ -17,20 +17,27 @@
...
       case _: EmptyTuple => Nil
-      case _: (t *: ts)  => summonInline[Equal[t]] :: summonAll[ts]
+      case _: (t *: ts)  => summon[A0, t] :: summonAll[A0, ts]
+    }
+
+  private inline def summon[A0, A]: Equal[A] =
+    compiletime.summonFrom {
+      case _: (A =:= A0) => compiletime.error("Cannot derive Equal.")
+      case _: (A <:< A0) => Equal.derived(using summonInline[Mirror.Of[A]])
+      case equal: Equal[A] => equal
```
Comment (bishabosha) [inline core/shared/src/main/scala-3/zio/prelude/EqualVersionSpecific.scala:40]: We should definitely add this pattern to the Scala 3 documentation, was there a specific example this helped with?

Comment (adamgfraser) [reply core/shared/src/main/scala-3/zio/prelude/EqualVersionSpecific.scala:40]: The derivation for `Color`. Would be great if we could push more of this complexity into the derivation functionality in Scala 3 so users did not have to deal with it.

Comment (sideeffffect) [reply core/shared/src/main/scala-3/zio/prelude/EqualVersionSpecific.scala:40]: Unfortunately it doesn't seem to work for recursive types like `List`.

https://scastie.scala-lang.org/y8VjOMW2TtuLsu2Jd9fIPA

On the other hand, @bishabosha 's solution from https://github.com/lampepfl/dotty/pull/16919#issuecomment-1431257814 works for `List`-like types too.

https://scastie.scala-lang.org/rq13aZFWQkq4xZA1sd79MA

Comment (adamgfraser) [reply core/shared/src/main/scala-3/zio/prelude/EqualVersionSpecific.scala:40]: Will take a look.

Comment (bishabosha) [reply core/shared/src/main/scala-3/zio/prelude/EqualVersionSpecific.scala:40]: > The derivation for `Color`. Would be great if we could push more of this complexity into the derivation functionality in Scala 3 so users did not have to deal with it.

Maybe, the idea was that Mirror + derived is intentionally the minimal possible implementation, and things like magnolia, shapeless, etc made a higher level api on top of that

Comment (bishabosha) [reply core/shared/src/main/scala-3/zio/prelude/EqualVersionSpecific.scala:40]: but we can look at defining more helpers for sure

**Pull request merged**